### DISCUSSION TITLE
Writing only to public schema with Document Hierarchy

### DISCUSSION THREAD

**Discussion by kamil-korus** [JasperFx/marten#2300] [General]
Hello. I have the following test classes:

```
   public abstract class NodeModel
    {
        //public string Id => Name;
        public Guid Id { get; } = Guid.NewGuid();
        public string Name { get; protected set; }        
    }

    public class NodeModel1 : NodeModel
    {
        public NodeModel1()
        {
            Name = "Node model 1";
        }        
    }

    public class NodeModel2 : NodeModel
    {
        public NodeModel2()
        {
            Name = "Node model 2";
        }
    }

    public  class TestModel
    {
        public Guid Id { get; } = Guid.NewGuid();
        public string Name => "Test model";
    }
```
I want to store NodeModel1 and NodeModel2 in one table using Document Hierarchies. Additionally, I intend to store the classes in different (already created) schemas, not the public one. Here is part of my repository:

```
public class MartenDocumentRepository<T> : IDocumentRepository<T> 
    {
        private readonly IDocumentStore documentStore;
        public MartenDocumentRepository(IDocumentStore documentStore)
        {
            this.documentStore = documentStore;
        }

        public async Task Add(T element, string schemaName)
        {            
            documentStore.Options.DatabaseSchemaName = schemaName;
            var documentSession =  documentStore.LightweightSession();

            documentSession.Store(element);
            await documentSession.SaveChangesAsync();
        }
...
```

When I tried to store NodeModel1 and NodeModel2 classes with no configuration, I got an error:
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'Unknown subclass type 'EasyWiseVP.Nodes.NodeModel1' for Document Hierarchy EasyWiseVP.Nodes.NodeModel')
So I added the following configuration:
```
builder.Services.AddMarten(options =>
{
    options.Connection(builder.Configuration.GetConnectionString("Marten"));
    **options.Schema.For<NodeModel>().AddSubClassHierarchy();**
});
```
Now, I don't have the error, but the classes are always stored in the public schema. When I try it for TestModel class (with no hierarchy), changing the schema works.

What can I do to alter between schemas also having the hierarchies? Thank you.

**Comment by jeremydmiller** [comment]
You need to specify the database schema name **only** at configuration time. You cannot do that at runtime, or at least that will not be honored. Try this:

```
builder.Services.AddMarten(options =>
{
    options.Connection(builder.Configuration.GetConnectionString("Marten"));
    **options.Schema.For<NodeModel>().DatabaseSchemaName("your schema name").AddSubClassHierarchy();**
});
```

**Comment by kamil-korus** [comment]
@jeremydmiller Thanks for the answer. But this is my case where I think it would be beneficial: I have a master entity (Container) with lists of "sub entities" (Nodes). Now, I want to store every Container in its own schema (created dynamically - it works already). NodeModels (non-event documents) logically are bound with Containers, so I wanted to store these documents also in the dynamically created schema.
Or is it a bad idea for some reason?
